net.rocketpool/image/src/app.ts

```diff
@@ -19,145 +23,119 @@ interface Event {
   logIndex: number;
 }
 
+type Provider = ethers.providers.JsonRpcProvider | ethers.providers.WebSocketProvider;
+
 /////////////////////////////////////////////////////////////////////////////////////////
 
 async function readEventsFromETH(
-  startBlock,
-  endBlock,
-  clb: (_: Event) => boolean,
-  stride = 10000
-) {
+  rethContract: RethAbi,
+  startBlock: number,
+  endBlock: number,
+  clb: (_: EventRow) => boolean,
+): Promise<void> {
   console.log(
-    "Considering events in block range [%s, %s] (%s blocks)",
-    startBlock,
-    endBlock,
-    endBlock - startBlock
+    `Considering events in block range [${startBlock}, ${endBlock}] (${endBlock - startBlock})`,
   );
 
-  let fromBlock = startBlock;
-  let toBlock = fromBlock;
-
-  do {
-    fromBlock = toBlock + 1;
-    toBlock = Math.min(endBlock, fromBlock + stride);
+  const eventFilters = [
+    rethContract.filters.TokensMinted(),
+    rethContract.filters.TokensBurned()
+  ] as const;
+  const eventsRowsPromises = eventFilters.map(async (eventFilter) => {
+    const events = await rethContract.queryFilter(eventFilter, startBlock, endBlock);
+
+    return events.map(async (event) => {
+      const block = await rethContract.provider.getBlock(event.blockNumber);
+
+      return {
+        eventName: event.event,
+        amount: event.args.amount.toString(),
+        ethAmount: event.args.ethAmount.toString(),
+        blockNumber: event.blockNumber,
+        blockHash: event.blockHash,
+        blockTime: block.timestamp,
+        transactionIndex: event.transactionIndex,
+        transactionHash: event.transactionHash,
+        logIndex: event.logIndex,
+      } as EventRow;
+    });
+  });
+  // NOTE: Array<Promise<Array<Promise<EventRow>>>> -> Array<EventRow>
+  const eventRows = await Promise.all((await Promise.all(eventsRowsPromises)).flat());
```

Comment:
This looks very sketchy to me...

I'm trying to understand how many requests will we be making to the node here. It seems:
- for every block range `[startBlock, endBlock]`
  - for every event type `{TokensMinted, TokensBurned}`
    - for every block with event?

This seems highly inefficient.

The previous version made one request per block range. It was filtering by contract address, so we could get events we're not interested in, but I assume most of RETH contract events are mint/burn...

Reply:
> This seems highly inefficient.

On the contrary, since we query by specific event rather than filtering all events, it can be compared to SQL: full table scan vs index search (the current approach)

Querying (!) all `TokensMinted`, `TokensBurned` events using the test block (13325304 - 18861964) of the contract took me 11 seconds (without querying timestamps). 

With querying timestamps it takes long 52 seconds. And this is a large block range 18861964 - 13325304 = 5,536,660.

> It was filtering by contract address, so we could get events we're not interested in, but I assume most of RETH contract events are mint/burn...

I also thought so at first, but as soon as I logged how many events we skip (roughly speaking ~90-90%), I realized that our desired events are absolute minority: out of 5 events the most frequent are `Approval`, `Transfer`.

Reply:
@sergiimk

Reply:
> for every block with event?

**UPD:** not actual after https://github.com/kamu-data/kamu-contrib/pull/13#discussion_r1436059719